### PR TITLE
feat: add error message when connection fails due to IP access issues COMPASS-6842

### PR DESCRIPTION
**Pull request by alenakhineika**
Detect by URI if a user tries to connect to Atlas. If errors, overwrite the error message to help users troubleshoot the issue and point to the fact, that their IP might not be whitelisted in the Atlas network settings.

### CODE REVIEW

File: src/connect.ts [hunk 1]
```diff
@@ -17,6 +17,8 @@ import { StateShareClient, StateShareServer } from './ipc-rpc-state-share';
 import ConnectionString, { CommaAndColonSeparatedRecord } from 'mongodb-connection-string-url';
 import EventEmitter from 'events';
 
+const isAtlas = (str: string) => str.match(/mongodb.net[:/]/i);
```
Comment (addaleax) [inline src/connect.ts]: Let’s maybe make this a bit more precise so that we don’t match `mongodb://supermongo.net/` or `mongodb://localhost/mongodb.net:foo`?

```suggestion
function isAtlas(str: string): boolean {
  try {
    const { hosts } = new ConnectionString(str);
    return hosts.every(host => /(^|\.)mongodb.net(:|$)/.test(host));
  } catch { return false; }
}
```

(not tested!)

File: src/connect.ts [hunk 2]
```diff
@@ -76,6 +78,8 @@ async function connectWithFailFast(uri: string, client: MongoClient, logger: Con
     if (failEarlyClosePromise !== null) {
       await failEarlyClosePromise;
       throw failedConnections.values().next().value; // Just use the first failure.
+    } else if (isAtlas(uri)) {
```
Comment (addaleax) [inline src/connect.ts]: Why `else`?

File: src/connect.ts [hunk 3]
```diff
@@ -76,6 +78,8 @@ async function connectWithFailFast(uri: string, client: MongoClient, logger: Con
...
+      (err as Error).message = 'It looks like this is a MongoDB Atlas cluster. Please ensure that your IP whitelist allows connections from your network.';
```
Comment (addaleax) [inline src/connect.ts]: Why are we replacing the entire error message here, and why are we not checking what type of error we saw?

Comment (addaleax) [reply src/connect.ts]: Also … I’m not seeing the phrase “whitelist” in the Atlas UI, only “Network Access” or “IP Access List”. I assume that we’re avoiding the term intentionally as described in e.g. [INIT-84](https://jira.mongodb.org/browse/INIT-84),  so let’s also not introduce it here.

File: src/ipc-rpc-state-share.ts [hunk 4]
```diff
@@ -109,7 +109,7 @@ export abstract class RpcServer {
 
       response = { status: 200, ...await this.handleRpc(content) };
     } catch (err) {
-      response = { status: 500, error: err && typeof err === 'object' && 'message' in err ? err.message : String(err) };
+      response = { status: 500, error: err && typeof err === 'object' && 'message' in err ? (err as Error).message : String(err) };
```
Comment (addaleax) [inline src/ipc-rpc-state-share.ts]: Actually, why this change? We’re already explicitly checking whether `err.message` exists on this line

Comment (alenakhineika) [reply src/ipc-rpc-state-share.ts]: Otherwise TS complaints with `error TS2339: Property 'message' does not exist on type 'object'`

Comment (addaleax) [reply src/ipc-rpc-state-share.ts]: It doesn’t in CI and it doesn’t locally for me, there might just be something wrong with your local setup? (But even then, TS _should_ see the `'message' in err` check, which should prevent it from presenting this type of error message)

Comment (alenakhineika) [reply src/ipc-rpc-state-share.ts]: Tried to bump typescript package from 4.5.4 to 5.1.3 and the error is gone. I wonder what was your version when you tried locally.

File: src/connect.ts (hunk 2 above)
Comment (alenakhineika) [reply src/connect.ts]: I asked first :))

File: src/ipc-rpc-state-share.ts (hunk 4 above)
Comment (addaleax) [reply src/ipc-rpc-state-share.ts]: Mh yeah, we might want to bump the minimum `typescript` version in `package.json`, although you should usually get that anyway on a fresh checkout because we’re not checking in the `package-lock.json` file

File: src/connect.ts (hunk 3 above)
Comment (alenakhineika) [reply src/connect.ts]: Yeah, seems that according to the doc: `For Atlas “IP Whitelist” = “Network Access List”`

File: src/connect.spec.ts [hunk 5]
```diff
@@ -299,6 +299,19 @@ describe('devtools connect', () => {
     });
   });
 
+  describe('atlas', () => {
+    it('shows the ensure that your network access list allows connections from your ip error', async() => {
```
Comment (addaleax) [inline src/connect.spec.ts]:
```suggestion
    it('shows the "ensure that your network access list allows connections from your ip" error', async() => {
```

or

```suggestion
    it('shows an error reminding the user to ensure that their network access list allows connections from their ip', async() => {
```

File: src/connect.ts [hunk 6]
```diff
@@ -72,11 +79,14 @@ async function connectWithFailFast(uri: string, client: MongoClient, logger: Con
...
   try {
     await client.connect();
-  } catch (err: unknown) {
+  } catch (err: any) {
     if (failEarlyClosePromise !== null) {
       await failEarlyClosePromise;
       throw failedConnections.values().next().value; // Just use the first failure.
     }
+    if (err.constructor?.name === 'MongoServerSelectionError' && isAtlas(uri)) {
+      err.message = `${err.message} It looks like this is a MongoDB Atlas cluster. Please ensure that your Network Access List allows connections from your IP.`;
```
Comment (addaleax) [inline src/connect.ts]: This might read a bit odd if `err.message` does not end in a `.`

File: src/connect.ts (hunk 2 above)
Comment (addaleax) [reply src/connect.ts]: This isn’t really resolved, right?

File: src/connect.ts [hunk 7]
```diff
@@ -73,11 +80,15 @@ async function connectWithFailFast(uri: string, client: MongoClient, logger: Con
...
     await client.connect();
   } catch (err: unknown) {
+    let connectErr = err;
     if (failEarlyClosePromise !== null) {
       await failEarlyClosePromise;
-      throw failedConnections.values().next().value; // Just use the first failure.
+      connectErr = failedConnections.values().next().value; // Just use the first failure.
+    }
+    if ((connectErr as any)?.constructor?.name === 'MongoServerSelectionError' && isAtlas(uri)) {
+      (connectErr as Error).message = `${(connectErr as Error).message}. It looks like this is a MongoDB Atlas cluster. Please ensure that your Network Access List allows connections from your IP.`;
```
Comment (addaleax) [inline src/connect.ts]: [not blocking]

I think there might still be some confusion here around the `catch (err: unknown)` vs `catch (err: any)` …

`err: unknown` is the best match when we’re expecting an error being thrown from code outside of our control, which might realistically throw values that aren’t `Error` instances, where we want to force ourselves to carefully do typechecking in TS with `typeof err` and `'...' in err`.

`err: any` is reasonable when we can assume that `err` is always an `Error` object, and we don’t really need to do manual typechecking in the places where we access it.

But here we’re mixing the two, we’re using `err: unknown` but then later basically assuming that `err` is an `Error` object anyway. That’s okay, but you could as well just say `err: any` in the first place then.

**Review by addaleax — APPROVED** [review]
